com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/CompletionsProvider.java

```diff
@@ -0,0 +1,105 @@
+        this.context = context;
+    }
+
+    @Override
+    public List<CompletionItem> codeComplete(StackFrame frame, String snippet, int line, int column) {
+        List<CompletionItem> res = new ArrayList<CompletionItem>();
+
+        try {
+            IType type = resolveType(frame);
+            if (type != null) {
+                final int offset = JsonRpcHelpers.toOffset(type.getCompilationUnit().getBuffer(), frame.location().lineNumber(), 0);
+                CompletionProposalRequestor collector = new CompletionProposalRequestor(type.getCompilationUnit(), offset);
+
+                collector.setAllowsRequiredProposals(CompletionProposal.FIELD_REF, CompletionProposal.TYPE_REF, true);
+                collector.setAllowsRequiredProposals(CompletionProposal.FIELD_REF, CompletionProposal.TYPE_IMPORT, true);
+                collector.setAllowsRequiredProposals(CompletionProposal.FIELD_REF, CompletionProposal.FIELD_IMPORT, true);
+
+                collector.setAllowsRequiredProposals(CompletionProposal.METHOD_REF, CompletionProposal.TYPE_REF, true);
+                collector.setAllowsRequiredProposals(CompletionProposal.METHOD_REF, CompletionProposal.TYPE_IMPORT, true);
+                collector.setAllowsRequiredProposals(CompletionProposal.METHOD_REF, CompletionProposal.METHOD_IMPORT, true);
+
+                collector.setAllowsRequiredProposals(CompletionProposal.CONSTRUCTOR_INVOCATION, CompletionProposal.TYPE_REF, true);
+
+                collector.setAllowsRequiredProposals(CompletionProposal.ANONYMOUS_CLASS_CONSTRUCTOR_INVOCATION, CompletionProposal.TYPE_REF, true);
+                collector.setAllowsRequiredProposals(CompletionProposal.ANONYMOUS_CLASS_DECLARATION, CompletionProposal.TYPE_REF, true);
+
+                collector.setAllowsRequiredProposals(CompletionProposal.TYPE_REF, CompletionProposal.TYPE_REF, true);
+
+                type.codeComplete(snippet.toCharArray(), offset, snippet.length(), null, null, null, frame.location().method().isStatic(), collector);
+
+                List<org.eclipse.lsp4j.CompletionItem> items = collector.getCompletionItems();
+
+                if (items != null) {
+                    for (org.eclipse.lsp4j.CompletionItem lspItem : items) {
+                        res.add(convertFromLsp(lspItem));
+                    }
+                }
+            }
+
+        } catch (DebugException | CoreException e) {
+            logger.log(Level.WARNING, String.format("Failed to code complete because of %s", e.toString()), e);
+        }
+
+        return res;
+    }
+
+    private IType resolveType(StackFrame frame) throws CoreException, DebugException {
+        ISourceLookUpProvider sourceProvider = context.getProvider(ISourceLookUpProvider.class);
+        if (sourceProvider instanceof JdtSourceLookUpProvider) {
+            IJavaProject project = JdtUtils.findProject(frame, ((JdtSourceLookUpProvider) sourceProvider).getSourceContainers());
```

Comment:
Need consider the case where cannot find the project for the frame.

Reply:
Fixed.